### PR TITLE
Support MinGW cross-compilation to Win32 target

### PR DESCRIPTION
**Pull request by stukowski**
While creating a Windows build using the MinGW toolchain was possible before, this pull request makes the necessary changes to also support cross-compiling the KIM API for Win32 under Linux using the MinGW toolchain. Support for cross-compilation is useful, because Axel Kohlmeyer also [generates the official LAMMPS executables for Windows under Fedora Linux using the MinGW-w64 cross-compilation toolchain](http://packages.lammps.org/windows.html). Thus, the following changes will make it possible to include the KIM module in the official LAMMPS for Windows binary package in the future.

- Unlike MSYS2/MinGW on Windows, the Ubuntu Linux/MinGW cross-compilation environment does not readily provide the [dlfcn-win32 portability layer](https://github.com/dlfcn-win32/dlfcn-win32). In order to relieve the user from installing this dependency by hand, this pull request removes calls to ``dlopen()``/``dlsym()``/``dlclose()`` and replaces them with corresponding calls to native Win32 API functions ``LoadLibraryEx()``/``GetProcAddress()``/``FreeLibrary()`` such that the ``<dlfcn>`` portability layer is no longer needed.
- The Linux/MinGW cross-compilation toolchain includes gcc 7.3, which is an older compiler version providing C++17 support but no full ``std::filesystem`` library implementation. It seems that at least gcc 9 is needed for successfully building a codebase using the ``std::filesystem`` API with MinGW. Only the MSYS2/MinGW environment provides such a more recent gcc version. In order to support cross-compiling the KIM API under Ubuntu/MinGW, I've included the [``ghc::filesystem`` library](https://github.com/gulrak/filesystem) in this pull request (MIT license). It is a single-header drop-in replacement for the ``std::filesystem`` library and will be used by the KIM API on C++17 platforms without full STL filesystem support. In the future, once the Linux/MinGW cross compilation environment provides gcc 9 or newer, the drop-in replacement may be removed again from the KIM API codebase.

### CODE REVIEW

File: cpp/src/CMakeLists.txt [hunk 1]
```diff
@@ -107,11 +107,19 @@ target_include_directories(kim-loader PUBLIC
...
-  set_property(TARGET kim-api PROPERTY CXX_STANDARD 17)
-  set_property(TARGET kim-api PROPERTY CXX_STANDARD_REQUIRED YES)
-  set_property(TARGET kim-loader PROPERTY CXX_STANDARD 17)
-  set_property(TARGET kim-loader PROPERTY CXX_STANDARD_REQUIRED YES)
+  # Activate C++17 language standard on Win32 platform, because kim-api code uses C++ filesystem library internally.
+  # Note: The 'cxx_std_17' target compile feature was added by CMake version 3.8. 
+  # For older CMake versions, fall back to setting the CXX_STANDARD target property instead.
+  get_property(cxx_features GLOBAL PROPERTY CMAKE_CXX_KNOWN_FEATURES) 
+  if("cxx_std_17" IN_LIST cxx_features)
+    target_compile_features(kim-api PUBLIC cxx_std_17)
```
Comment (ellio167) [inline cpp/src/CMakeLists.txt:115]: What about setting these only for the `KIM_FILESYSTEM_Path.cpp` file?  Would this be an advantage/disadvantage?  I'm interested in your experience/views on "best practice" in this regard....

File: cpp/src/ghc/filesystem.hpp [hunk 2]
```diff
@@ -0,0 +1,5526 @@
+//---------------------------------------------------------------------------------------
```
Comment (ellio167) [inline cpp/src/ghc/filesystem.hpp:1]: I would suggest we name it `cpp/src/ghc-filesystem/`

Add this file to `.clang-format-hook-ignore` so that it is not reformatted by accident

Also, add a `ghc-filesystem.origin` file similar to `cpp/src/edn-cpp/edn-cpp.origin` (maybe use the release tag instead of or in addition to commit hash)

Comment (stukowski) [reply cpp/src/ghc/filesystem.hpp]: Done.

File: cpp/src/CMakeLists.txt (hunk 1 above)
Comment (stukowski) [reply cpp/src/CMakeLists.txt:115]: No, compiling only some .cpp files in C++17 mode and others in C++98 mode will likely lead to problems, because the two parts will not be ABI-compatible. In fact, I was struggling with this problem myself for quite some time last week. The ``collections-info`` executable crashed with a segmentation fault, because it was compiled in C++98 mode while the ``kim-loader`` library was compiled in C++17 mode (on Win32). At build time everything seemed to be fine (no compiler or linker warnings) but at runtime the program crashed, because ``sizeof(KIM::FILESYSTEM::Path)`` was not the same in different parts of the same executable. 
This is why I added ``target_compile_features(kim-loader PUBLIC cxx_std_17)`` in the CMake script, because it makes sure that the C++17 requirement will be passed on from the ``kim-loader`` target to other targets consuming it. This is not the case when just setting the ``CXX_STANDARD`` target property to 17.

File: cpp/src/KIM_FilesystemPath.cpp [hunk 3]
```diff
@@ -33,6 +33,7 @@
 
 #include <cstdlib>
 #include <iostream>
+#include <sstream>
```
Comment (ellio167) [inline cpp/src/KIM_FilesystemPath.cpp:36]: Why is this needed?  It seems to work without it.

File: cpp/src/CMakeLists.txt (hunk 1 above)
Comment (ellio167) [reply cpp/src/CMakeLists.txt:115]: Nice!  Great to know. Thanks for the explanation.

File: cpp/src/KIM_FilesystemPath.cpp (hunk 3 above)
Comment (stukowski) [reply cpp/src/KIM_FilesystemPath.cpp:36]: It's needed for [``std::basic_istringstream``](https://en.cppreference.com/w/cpp/io/basic_istringstream) referenced in ``PathList::Parse()``. Yeah, on most platforms you can get away without an explicit ``<sstream>`` include, but Linux/MinGW requires it.

Comment (ellio167) [reply cpp/src/KIM_FilesystemPath.cpp:36]: OK, please add something like `// For Linux/MinGW  IWYU pragma: keep`

Comment (stukowski) [reply cpp/src/KIM_FilesystemPath.cpp]: Done
